lib/go/thrift/simple_server.go

```diff
@@ -187,12 +187,24 @@ func (p *TSimpleServer) processRequests(client TTransport) error {
 	if err != nil {
 		return err
 	}
-	outputTransport, err := p.outputTransportFactory.GetTransport(client)
-	if err != nil {
-		return err
-	}
 	inputProtocol := p.inputProtocolFactory.GetProtocol(inputTransport)
-	outputProtocol := p.outputProtocolFactory.GetProtocol(outputTransport)
+	var outputTransport TTransport
+	var outputProtocol TProtocol
+
+	// for THeaderProtocol, we must use the same protocol instance for
+	// input and output so that the response is in the same dialect that
+	// the server detected the request was in.
+	if _, ok := inputProtocol.(*THeaderProtocol); ok {
```

Comment:
Hmm, interesting, usually the server layer has no special cases for the protocol.

Reply:
Without this logic the outputProtocol won't know it should use a non-default protocol detected on inputProtocol.

- python: https://github.com/apache/thrift/blob/823474ec89355f72d3f0720ae5dacc2036d41c03/lib/py/src/server/TServer.py#L189-L191
- cpp: https://github.com/apache/thrift/blob/823474ec89355f72d3f0720ae5dacc2036d41c03/lib/cpp/src/thrift/server/TNonblockingServer.cpp#L395-L403